### PR TITLE
[CDEC-509] Remove HasCoreConfiguration and HasConfiguration

### PR DESCRIPTION
**Pull request by ruhatch**
## Description

This is the removal of the final `Data.Reflection` constraint in `core`! This PR removes the `dbSerializeVersion` configuration parameter, a so far unused value. Could reviewers please check that a node running the code before this commit **continues to sync after this commit is applied**.

## Linked issue

[CDEC-509](https://iohk.myjetbrains.com/youtrack/issue/CDEC-509)

## Type of change

- [~] 🐞 Bug fix (non-breaking change which fixes an issue)
- [~] 🛠 New feature (non-breaking change which adds functionality)
- [~] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [~] 🔨 New or improved tests for existing code
- [~] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [~] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [~] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## QA Steps

1) Connect to the testnet running a node on `develop` and sync some blocks
2) Kill the node and then with **the same state** connect with a node from this PR
3) Verify that the DB is still read and the syncing continues

### CODE REVIEW

**Comment by erikd** [comment]
Probably worth rebasing that against current develop (its 38 commits behiind) to reduce the chances of merge errors when it is merged to develop.

File: db/src/Pos/DB/Functions.hs [hunk 1]
```diff
@@ -7,91 +7,40 @@ module Pos.DB.Functions
...
+    dbGet tag key >>= traverse (either throwM pure . dbDecodeIgnoreVersion)
 
 -- | Write serializable value to DB for given key. Uses simple versioning.
 dbPutBi :: (Bi v, MonadDB m) => DBTag -> ByteString -> v -> m ()
-dbPutBi tag k v = dbPut tag k (dbSerializeValue v)
-
--- | Version of 'serialize'' function that includes version when serializing a value.
-dbSerializeValue :: (HasCoreConfiguration, Bi a) => a -> ByteString
-dbSerializeValue = serialize' . (dbSerializeVersion,)
+dbPutBi tag k v = dbPut tag k (serialize' v)
```
Comment (erikd) [inline db/src/Pos/DB/Functions.hs]: Hmm, want to check that a little more carefully on Monday.

Marking this "requires changes" until I properly examine the consequences of this.

Comment (erikd) [reply db/src/Pos/DB/Functions.hs]: The weird thing is that from looking at the code, I expect this to break when accessing existing DBs. The fact that it doesn't freaks me out even more than if it did break.

I know its a pain in the neck, but we can we separate out the `dbSerializeVersion` parts of this PR and maybe even have a test to show that its safe. The idea would be to:
* Raise a PR with some tests that use `dbSerializeVersion` as it is now.
* Once that PR is merged, raise a PR removing `dbSerializeVersion`.

Comment (ruhatch) [reply db/src/Pos/DB/Functions.hs]: You're right, that's safer and will reduce the complexity of this PR. I'm getting on it now.

Comment (ruhatch) [reply db/src/Pos/DB/Functions.hs]: This changes the DB to write without a version, but read using:
```
dbDecodeIgnoreVersion :: forall v . Bi v => ByteString -> Either DBError v
dbDecodeIgnoreVersion bytes = case decodeFull' @v bytes of
    Right val -> Right val
    Left _    -> bimap DBMalformed snd $ decodeFull' @(Word8, v) bytes
```
which first tries to decode without a version and, if that fails, tries again with the version and drops it

Comment (ruhatch) [reply db/src/Pos/DB/Functions.hs]: That's why it doesn't break on old DBs :)

Comment (ruhatch) [reply db/src/Pos/DB/Functions.hs]: The tests are in place and passing. Ready to re-review, paying particular attention to [this file](https://github.com/input-output-hk/cardano-sl/pull/3550/files#diff-1f5ccc5e83ce78d1e0ac185490714322)

**Comment by erikd** [comment]
Something wend *way* wrong with the changes to `lib/configuration.yam`.

Github reports:
> 14,646 additions, 14,676 deletions not shown because the diff is too large. Please use a local Git client to view these changes. 

I suspect the "semantic changes only" version if the diff for that file would be 30 lines. We really need *that* version :)

**Comment by erikd** [comment]
I compiled this version and ran it on an existing node DB (already synced to the 69th epoch or mainnet) and it ran fine.

I then deleted the the DB to force a full sync and that also seems to be fine and is currently at the 10th epoch.

**Pull request merged**